Fixing the thrusters' control calculation

rafaelsaback opened the pull request:
AccelerationController: Now this task computes the thrusters' control based on the SVD of the thruster matrix

BasePIDController: Added the function resetController

PIDController: Added an if statement in order to avoid a NaN return from normalized function when the angular velocity is a null vector.

Signed-off-by: Rafael Saback rafael.saback@fieb.org.br

doudou commented on auv_control.orogen:
```diff
@@ -99,6 +99,8 @@ task_context 'BasePIDController' do
     # The states of the pid-controllers
     output_port 'pid_state', 'base::LinearAngular6DPIDState'
+        
+    operation("resetController")
```
I you need to reset the controller, run a stop/cleanup/configure/start cycle

doudou replied:
actually stop/start should be enough

rafaelsaback replied:
Ok!

doudou commented on tasks/AccelerationController.cpp:
```diff
@@ -25,16 +25,17 @@ bool AccelerationController::configureHook()
     }
-    thrusterMatrix = _matrix.get().transpose();
+    thrusterMatrix = _matrix.get();
+    numberOfThrusters = thrusterMatrix.cols();
```
don't store numberOfThrusters as a class attribute, it duplicates information. If it clarifies the code in this method (as it seems), just cache it as local variable(s). Otherwise, you can also make it a method which would extract the information from the matrix.

rafaelsaback replied:
Ok!

doudou commented:
The original idea for this task was to provide a way to hand-tune a X/Y/Z controller for people that don't have access to some model-fitting method. The SVD change obviously changes this, as instead of having a straightforward acceleration-to-thruster values, we now are supposed to provide a thruster-to-acceleration mapping (which, I guess, is what comes from the model fitting)

I'm not against it per se, but (1) this is a pretty big change (under the wraps of a small code change) and (2) do we really want to get rid of the original implementation ? Can you tell how one can hand-tune  a controller with the change applied ?

rafaelsaback commented:
Hey @doudou , thanks for your comments, they've clarified some things to me and I do agree with the first ones.

I think I understood most of what you've said in the last comment, but let me make sure that we're walking over the same ground. The way I see the acceleration controller is what Fossen calls thruster mapping, and it basically tries to distribute the forces and moments that were calculated by the velocity controller between the thrusters. In other words, these forces and moments calculated by the velocity controller are expressed relative to the center of mass of the vehicle and the acceleration controller distributes these efforts for the thrusters by using the Thruster Configuration Matrix (TCM).

I know that finding a proper TCM might sound a bit ideal for real applications, but it is not. An approximation of this matrix can be found by taking the distance of the thrusters to the center o mass as well as its directions. A hand tuning of the TCM can also be performed, and in my personal opinion, it's easier to tune the TCM because it's based on geometric properties (thruster direction and position). The previews way can be more straightforward for the user who is used to this method, but, since it's not grounded in any theory nor in physical parameters, it might sound a bit confuse and unreasonable for a new user.

Also, the old way do needs to perform real experiments with the vehicle, while with this new way you can get an approximation of the TCM just by taking the geometric position of the thrusters and considering the center of mass coincident with the body frame. For a initial and reasonable approach you don't need to deal with the logistics of putting the vehicle in water. And if we have a well developed project, like in Flatfish case, we can find the center of mass of the vehicle as well as the relative position of the thrusters using SolidWorks.

Well, this is pretty much what I think about these two methods, but if you want an example to ilustrate this, I present one in the attachment that shows these two different ways of calculating the thruster's efforts, and I also present some references and part of the theory behind this new solution.

![example_acc_controller](https://cloud.githubusercontent.com/assets/7912190/6916195/81e6ee00-d79a-11e4-94cf-2c5b9110c627.png)

Link referenced in the example: https://inst.eecs.berkeley.edu/~ee127a/book/login/def_pseudo_inv.html

ViVittori commented:
Hi Rafael , Sylvain 

Thank you for your explanation Rafael. 

Basically if I understand correctly you are saying that the command vector was not correctly calculated until now ?! 

I took  the geometric position of the thrusters for the TCM  out of our solidworks model and it "worked" on our simulator afterwards. 
I thought that it was pretty much straightforward to take the geometric position for the TCM.  
What would you have used instead in the "old" way Sylvain ? 

Then to me it is pretty obvious that it should be integrated in the code if it provides a better result.

Cheers

V.v

doudou commented:
Thanks a lot for the long explanation Rafael. I think I understand it better now.

You _can_ actually hand-tune the matrix we had before using geometry, by filling it with the relative contributions of each thruster on a particular axis. However, I do understand and agree that this method of building the thruster-to-force matrix is more grounded in the underlying mechanics.

We'll have to have the opinion of @saarnold before breaking backward compatibility like this, though. (Worst case, adding a flag for backward compatibility would not be that much work)

doudou commented:
@rafaelsaback: could you add the description of how to get an initial thruster matrix in the README ?

rafaelsaback commented:
@doudou, I've modified the README for explaining how one could obtain a first tuning of the thruster matrix. I've also added a .png example file in order to help the user understaning.

If you have further suggestions just let me know.

@ViVittori, with the previous calculation you could get a reasonable performance when you have a closed-loop control, but with the pseudo-inverse we can have an even better performance.

saarnold commented:
Im fine with merging it to master, but I think having a flag for backward compatibility would be good. The default would be the new version.

doudou commented:
:+1:

Commit:
AccelerationController: Now this task computes the thrusters' control based on the SVD of the thruster matrix

PIDController: Added an if statement in order to avoid a NaN return from normalized function when the angular velocity is a null vector.

rafaelsaback commented:
I've created a flag for the SVD method and set it as default. I've also moved the decomposition to the configureHook, and during the updateHook only the solve function will be executed.

doudou commented:
For the sake of completeness: I believe there is a major problem with using this method in the general case: saturation. 

A practical example: one has, say two non-centered "forward" thrusters and two non-centered side thrusters. As designers, we know that the forward thrusters should mainly be used for forward movement and the side thrusters for Y and Yaw movements. However, there's no constraint to tell that to the solver and it might decide otherwise, hitting saturation.

Other problem would happen if you had two small thrusters for pitch and a main dive thruster for Z. The solver might decide to use the pitch thrusters for descent, saturating them and removing the pitch control capability. 

I'm looking into creating a syskit-based UI for calibration of the control loop. For now, I'm going for picking the thruster positions from the SDF files, building the matrix as you describe but doing the inversion there (offline) and letting the user hand-tune the resulting matrix. If you have ideas on how to solve this problem (or whether it is a false problem that is just in my head ;-)), please let me know.

rafaelsaback commented:
Hey @doudou, you're correct. Whenever you have the possibility of controlling a specific DOF in different manners, if you use a pure and simple TCM  with the SVD solution you will not be able to determine which thruster should control each DOF.

The solution for this issue  is something that Fossen (1994, p. 98) has addressed before, which is called optimal distribution of propulsion and control forces. In a short way, in order to determine how to distribute the forces between the thrusters, you should also use a weighting cost matrix for the thrusters. So, if you consider a greater weighting cost for the surge thrusters, while controlling yaw, the sway thrusters will be used more instead, since the surge thrusters have a greater weighting cost than the sway thrusters (the example at the end makes this a bit clearer).

Now a bit longer explanation. Considering that we have more than one way of controlling some degrees of freedom, what should be calculated is the solution of an optimal problem that takes into account the following cost function:

![image](https://cloud.githubusercontent.com/assets/7912190/7367034/ef952558-ed9c-11e4-8242-2fed85da5457.png)

That should be minimized subject to:

![image](https://cloud.githubusercontent.com/assets/7912190/7367047/09131ea4-ed9d-11e4-9479-d04d7d8c1a27.png)

Where W is the wheighting cost matrix (square, positive definite and diagonal), B is the TCM and u is the force in each thruster. The solution of this problem will provide us with the inverse matrix that should be used to calculate the thrusters commands.

The demonstration of the solution can be seen in Fossen (1994, p. 98) but the resulting inverse is the following one:

![image](https://cloud.githubusercontent.com/assets/7912190/7367451/d51648b2-ed9f-11e4-9e00-6b2aa7363a2c.png)

If we make the matrix W equal to the identity, then the solution of this problem will be the previous mentioned right pseudo-inverse.

In terms of implementation in the task, it would be necessary to create a new property for the wheighting matrix. The calculation of the above inverse should be implemented in the task (it could be done in the configureHook), and then it should be multiplied by the component input. This requires a bit more of programming than the previous version, but it can be done without much problems. Let me know how it sounds to you guys.

We can see how it works based on the previous example you've mentioned:

PS: The example I show here can be done by using octave. But be aware to use "pinv" instead of "inv" when inverting the terms inside the parenthesis.

![image](https://cloud.githubusercontent.com/assets/7912190/7370279/467fbe7a-edb9-11e4-9926-7b3c2f1e2423.png)

![image](https://cloud.githubusercontent.com/assets/7912190/7370420/4b6a27e4-edba-11e4-8ff5-5eecbbc23eea.png)

![image](https://cloud.githubusercontent.com/assets/7912190/7370429/573a222c-edba-11e4-85cf-a6c5b7e67ca9.png)

doudou commented:
@rafaelsaback thanks a lot for the explanation !

saarnold commented:
This was successfully tested now for a while and can therefore be mergerd.
